drivers/timer/intel_adsp_timer.c

```diff
@@ -214,16 +220,24 @@ void smp_timer_init(void)
 }
 
 /* Runs on core 0 only */
-static int sys_clock_driver_init(void)
+static int sys_clock_driver_init(const struct device *dev)
 {
 	uint64_t curr = count();
-
-	IRQ_CONNECT(TIMER_IRQ, 0, compare_isr, 0, 0);
+	if (!sys_clock_driver_reinit)
+		IRQ_CONNECT(TIMER_IRQ, 0, compare_isr, 0, 0);
 	set_compare(curr + CYC_PER_TICK);
 	last_count = curr;
 	irq_init();
 	return 0;
 }
 
-SYS_INIT(sys_clock_driver_init, PRE_KERNEL_2,
-	 CONFIG_SYSTEM_CLOCK_INIT_PRIORITY);
+#ifdef CONFIG_PM
+void sys_clock_idle_exit(void)
+{
+	sys_clock_driver_reinit = true;
+	sys_clock_driver_init(DEVICE_DT_INST_GET(0));
+}
+#endif /* CONFIG_PM */
+
+DEVICE_DT_INST_DEFINE(0, sys_clock_driver_init, NULL, NULL, NULL,
```

Comment:
Why have you change it ? That makes this driver different from others sys_clock*. That is a pretty weird driver that and uses SYS_INIT and has some peculiarities as far as I remember. This change does not seem necessary since you don't need the device pointer for anything. Keep it simple and don't change it please.

Reply:
I do not change things that are not mandatory.
We need Device instance pointer to pass it to the clock init function. Previously this function was statically initialized with macros, now I need to call it run-time with proper argument.

Reply:
```
static int sys_clock_driver_init(const struct device *dev)
{
	uint64_t curr = count();
	if (!sys_clock_driver_reinit)
		IRQ_CONNECT(TIMER_IRQ, 0, compare_isr, 0, 0);
	set_compare(curr + CYC_PER_TICK);
	last_count = curr;
	irq_init();
	return 0;
}
```

Where `dev` is used there ? You can simply call `sys_clock_driver_init(NULL)`.

Reply:
Yes... Indeed I can. Call will look a bit weird but good catch. I will fix it tomorrow, thanks.

Reply:
It is better than this be the only one doing it IMHO :)